book/src/04_4_1_interrupts.md

```diff
@@ -12,77 +11,18 @@ cargo run --example solution
-
-2. Write the configuration into the register with [`unsafe extern "C" fn gpio_config`](https://esp-rs.github.io/esp-idf-sys/esp_idf_sys/fn.gpio_config.html). This needs to happen in the unsafe block. To make these FFI calls, we can use the macro `esp!($Cfunktion)`.
-
-3. Install a generic GPIO interrupt handler with [`unsafe extern "C" fn gpio_install_isr_service`](https://esp-rs.github.io/esp-idf-sys/esp_idf_sys/fn.gpio_install_isr_service.html). This function takes `ESP_INTR_FLAG_IRAM` as argument.
-
-4. Create a `static mut` that holds the queue handle we are going to get from `xQueueGenericCreate`. This is a number that uniquely identifies one particular queue, as opposed to any of the other queues in our program. The queue storage itself if managed by the Operating System.
-
-    ```rust
-    static mut EVENT_QUEUE: Option<QueueHandle_t> = None;
-    ```
-
-5. Create the event queue using [`pub unsafe extern "C" fn xQueueGenericCreate`](https://esp-rs.github.io/esp-idf-sys/esp_idf_sys/fn.xQueueGenericCreate.html). This lets us safely pass events from an interrupt routine to our main thread.
-
-    ```rust
-    EVENT_QUEUE = Some(xQueueGenericCreate(QUEUE_SIZE, ITEM_SIZE, QUEUE_TYPE_BASE));
-    ```
-
-6. Add a function which that will be called whenever there is a GPIO interrupt on our button pin. We put this function in a special block of RAM (`iram0`), so it will still be available even if the external flash is busy doing something else (like filesystem work). The function needs to get the queue handle from `EVENT_QUEUE` and call the `xQueueGiveFromISR` function with a `std::ptr::null_mut()` - the objects in our queue are of size zero, so we don't actually need a 'thing' to put on the queue. Instead, the act of pushing a 'nothing' is enough to wake up the other end!
-
-    ```rust
-    #[link_section = ".iram0.text"]
-    unsafe extern "C" fn button_interrupt(_: *mut c_void) {
-        xQueueGiveFromISR(EVENT_QUEUE.unwrap(), std::ptr::null_mut());
-    }
-    ```
-
-If the interrupt fires, an event is added to the queue.
-
-7. Pass the function we just wrote to the generic GPIO interrupt handler we registered earlier, along with the number of the GPIO pin that should cause this function to be executed.
-
-    ```rust
-    esp!(gpio_isr_handler_add(
-        GPIO_NUM,
-        Some(button_interrupt),
-        std::ptr::null_mut()
-    ))?;
-    ```
-
-8. Inside a loop, wait until the queue has an item in it. That is, until the `button_interrupt` function puts something in the queue.
-
-    ```rust
-    let res = xQueueReceive(EVENT_QUEUE.unwrap(), ptr::null_mut(), QUEUE_WAIT_TICKS);
-    ```
-
-9. Handle the value of `res`, so that "Button pushed!" is logged, if the button is pushed.
-
-10. Run the program and push the `BOOT` button, so see how it works!
+2. Instantiate a new notification and notifier
+    - See `hal::task::notification` documentation
+3. In an `unsafe` block, create a subscription and its callback function.
```

Comment:
We should try to educate what is unsafe about the subscribe mechanism, in regards to a) what it can run b) the closure capturing its environment and you can use static variables inserted in it ( like you can try to move a PinDriver inside the ISR context with it). All captured variables in any form need to outlive the subscription to be valid | even more when used with subscribe_nonstatic . Potential problems around  mem::forget etc -> From there we can go to maybe a simple async example where the user dont need all that. Pro: simpler to use / no unsafe for user Con: Doesn't directly run in ISR (only triggerd by it) which is ok for 99% of time